File: fel-adapter/src/de/proskor/fel/impl/EventTypeContainerImpl.scala
package de.proskor.fel.impl

import java.util.{List => JavaList}

import scala.collection.JavaConversions.seqAsJavaList

import de.proskor.automation.Element
import de.proskor.automation.Repository
import de.proskor.fel.container.EventInstanceContainer
import de.proskor.fel.container.EventTypeContainer
import de.proskor.fel.event.EventType

class EventTypeContainerImpl(val peer: Element) extends EntityImpl(peer) with EventTypeContainer {
  override def getEvents: JavaList[EventType] = {
    val events = for {
      connector <- peer.connectors
      if connector.target == peer && connector.stereotype == "belongsTo"
      source = connector.source
    } yield new EventTypeImpl(source)
    events.toList
  }

  override def getInstances: JavaList[EventInstanceContainer] = {
    val instances = for {
      connector <- peer.connectors
      if connector.target == peer && connector.stereotype == "instanceOf"
      source = connector.source
    } yield new EventInstanceContainerImpl(source)
    instances.toList
  }

  override def createEvent(name: String): EventType = {
    val repository = Repository.instance
    val model = repository.models.headOption getOrElse repository.models.add("Model", "Package")
    val fel = model.packages.find(_.name == "FEL") getOrElse model.packages.add("FEL", "Package")
    val event = fel.elements.add(name, "Object")
    event.stereotype = "EventType"
    val connector = event.connectors.add("", "Connector"); connector.stereotype = "belongsTo"
    connector.source = event
    connector.target = peer
    new EventTypeImpl(event)
  }

  override def addInstance(instance: EventInstanceContainer) {}

  override def getChildren: JavaList[EventTypeContainer] = List() // TODO
  override def getParent: EventTypeContainer = null // TODO
}

File: ea-fel/src/de/proskor/fel/stub/EventTypeContainerImpl.java
package de.proskor.fel.stub;

import java.util.ArrayList;
import java.util.List;

import de.proskor.fel.container.EventInstanceContainer;
import de.proskor.fel.container.EventTypeContainer;
import de.proskor.fel.event.EventType;

public class EventTypeContainerImpl extends EntityImpl implements EventTypeContainer {
	private List<EventType> eventTypes = new ArrayList<EventType>();
	private List<EventInstanceContainer> instances = new ArrayList<EventInstanceContainer>();
	private final EventTypeContainer parent;
	private ArrayList<EventTypeContainer> children = new ArrayList<EventTypeContainer>();

	public EventTypeContainerImpl() {
		parent = null; 
	}
	
	public EventTypeContainerImpl(EventTypeContainer parent) {
		this.parent = parent; 
		parent.getChildren().add(this);
	}
	
	@Override
	public List<EventType> getEvents() {
		return this.eventTypes;
	}

	@Override
	public List<EventInstanceContainer> getInstances() {
		return this.instances;
	}

	@Override
	public EventType createEvent(String name) {
		EventTypeImpl eventType = new EventTypeImpl();
		eventType.setName(name);
		eventType.setContainer(this);
		this.eventTypes.add(eventType);
		return eventType;
	}

	@Override
	public void addInstance(EventInstanceContainer instance) {
		this.instances.add(instance);
		((EventInstanceContainerImpl)instance).setType(this);
	}

	@Override
	public EventTypeContainer getParent() {
		return parent;
	}

	@Override
	public List<EventTypeContainer> getChildren() {
		return children;
	}
}
